File: exercises/5_saga/practice/src/main/java/fulfillment/FulfillmentActivities.java
package fulfillment;

import io.temporal.activity.ActivityInterface;
import io.temporal.activity.ActivityMethod;

@ActivityInterface
public interface FulfillmentActivities {

    @ActivityMethod
    String reserveInventory(Order order);

    @ActivityMethod
    String processPayment(Order order);

    @ActivityMethod
    String dispatchToFulfillment(Order order, String reservationId);

    // Compensating transactions — run in reverse order on failure
    @ActivityMethod
    void releaseInventory(String reservationId);

    @ActivityMethod
    void refundPayment(String paymentConfirmation);
}

File: exercises/2_child_workflows/solution/src/main/kotlin/fulfillment/FulfillmentActivities.kt
package fulfillment

import io.temporal.activity.ActivityInterface
import io.temporal.activity.ActivityMethod

@ActivityInterface
interface FulfillmentActivities {

    @ActivityMethod
    fun processPayment(order: Order): String

    @ActivityMethod
    fun dispatchToFulfillment(order: Order, reservationId: String): String
}
